torch/_dynamo/variables/torch.py

```diff
@@ -832,9 +832,25 @@ def get_comparable_state(state):
             )
 
         def speculate_subgraph(f, sub_args, graph_checkpoint, checkpoint):
-            # Setup the subgraph we're going to capture into
-            tx.output.graph = torch.fx.Graph()
-            tx.output.input_name_to_proxy.clear()
+            # NOTE: [The new speculate_subgraph strategy]
+            # We temporarily overwrite tx.output with a new OutputGraph that
+            # we capture f into and save the original OutputGraph as tx.base_output.
+            #
+            # We need to keep tx.base_output around because we still need to
+            # perform operations on it. Namely, if we encounter a new global
+            # while tracing f, then we need to lift that global to be an input
+            # of tx.base_output.
```

Comment:
Discussed offline before, but I think we should consider a design around giving OutputGraph parentage.

Then, you initialize your new OutputGraph like:

```
torch._dynamo.output_graph.OutputGraph(tx.output, ...
```

And then instead of

```
if self.base_output is not None:
```
And doing the argument lifting a such, you do (wrote psuedocode, most of the logic for _record below is in LOAD_GLOBALS):

```
def record(self, var, ...)
     self._record(var) # Do internal bookeeping for current output_graph
     self.parent.record(var, ...) # Pass it up a level
```

This would give you infinite nesting without needing to keep around a `base_output` right?